sp/src/game/shared/mapbase/vscript_funcs_shared.cpp

```diff
@@ -1006,11 +1467,507 @@ bool ScriptIsClient()
+{
+	int idx = GetIndexForContext(szId);
+
+	KeyValues *pKV;
+
+	if ( idx == -1 )
+	{
+		pKV = new KeyValues("ScriptSavedTable");
+		m_aKeyValues.AddToTail(pKV);
+
+		if ( V_strlen(szId) > 255 ) // arbitrary clamp
+		{
+			char c[256];
+			V_strncpy( c, szId, sizeof(c) );
+			m_aContext.CopyAndAddToTail(c);
+		}
+		else
+		{
+			m_aContext.CopyAndAddToTail(szId);
+		}
+	}
+	else
+	{
+		pKV = m_aKeyValues[idx];
+		pKV->Clear();
+	}
+
+	ScriptVariant_t key, val;
+	int nIterator = -1;
+	while ( ( nIterator = g_pScriptVM->GetKeyValue( hTable, nIterator, &key, &val ) ) != -1 )
+	{
+		switch ( val.m_type )
+		{
+			case FIELD_FLOAT:   pKV->SetFloat ( key.m_pszString, val.m_float     ); break;
+			case FIELD_INTEGER: pKV->SetInt   ( key.m_pszString, val.m_int       ); break;
+			case FIELD_BOOLEAN: pKV->SetBool  ( key.m_pszString, val.m_bool      ); break;
+			case FIELD_CSTRING: pKV->SetString( key.m_pszString, val.m_pszString ); break;
+		}
+
+		g_pScriptVM->ReleaseValue(key);
+		g_pScriptVM->ReleaseValue(val);
+	}
+}
+
+//-----------------------------------------------------------------------------
+// Retrieves a table from storage. Write into input table.
+//-----------------------------------------------------------------------------
+void CScriptSaveRestoreUtil::RestoreTable( const char *szId, HSCRIPT hTable )
+{
+	int idx = GetIndexForContext(szId);
+
+	KeyValues *pKV;
+
+	if ( idx == -1 )
+	{
+		// DevWarning( 2, "RestoreTable could not find saved table with context '%s'\n", szId );
+		return;
+	}
+	else
+	{
+		pKV = m_aKeyValues[idx];
+	}
+
+	FOR_EACH_SUBKEY( pKV, key )
+	{
+		switch ( key->GetDataType() )
+		{
+			case KeyValues::TYPE_STRING: g_pScriptVM->SetValue( hTable, key->GetName(), key->GetString() ); break;
+			case KeyValues::TYPE_INT:    g_pScriptVM->SetValue( hTable, key->GetName(), key->GetInt()    ); break;
+			case KeyValues::TYPE_FLOAT:  g_pScriptVM->SetValue( hTable, key->GetName(), key->GetFloat()  ); break;
+		}
+	}
+}
+
+//-----------------------------------------------------------------------------
+// Remove a saved table.
+//-----------------------------------------------------------------------------
+void CScriptSaveRestoreUtil::ClearSavedTable( const char *szId )
+{
+	int idx = GetIndexForContext(szId);
+
+	if ( idx == -1 )
+	{
+		// DevWarning( 2, "ClearSavedTable could not find saved table with context '%s'\n", szId );
+		return;
+	}
+
+	m_aKeyValues[idx]->deleteThis();
+	m_aKeyValues.Remove(idx);
+
+	delete[] m_aContext[idx];
+	m_aContext.Remove(idx);
+}
+
+//-----------------------------------------------------------------------------
+//
+//-----------------------------------------------------------------------------
+#define SCRIPT_MAX_FILE_READ_SIZE  (16 * 1024)			// 16KB
+#define SCRIPT_MAX_FILE_WRITE_SIZE (64 * 1024 * 1024)	// 64MB
+#define SCRIPT_RW_PATH_ID "MOD"
+#define SCRIPT_RW_FULL_PATH_FMT "export/%s"
+
+class CScriptReadWriteFile : public CAutoGameSystem
```

Comment:
Since it's not possible to exit to an upper directory, the only issue I can think of is writing very large files to fill up the disk. The same issue exists in L4D2 too. I'm assuming Valve thought about it enough to allow it. I added the already-too-large 64 MB write size because no limit in Valve's vscript seemed odd. I have not tested Dota 2, however, it may or may not have different rules.

For multiplayer it can be used to export data to players (clientside), instead of only saving on server; but the same issue also applies.

Binaries cannot be written or read; malicious batch/shell scripts can be written, but that still requires the user to manually execute them.